### PR TITLE
Refactor CMake, setup f3d SDK, and make VTK a private dependency

### PR DESCRIPTION
**Pull request by Meakk**

### CODE REVIEW

File: src/application/main.cxx [hunk 1]
```diff
@@ -9,10 +9,6 @@ int main(int argc, char** argv)
 
   try
   {
-#if NDEBUG
-    vtkObject::GlobalWarningDisplayOff();
-#endif
-
```
Comment (mwestphal) [inline src/application/main.cxx:15]: ?

File: src/library/CMakeLists.txt [hunk 2]
```diff
@@ -2,10 +2,7 @@
 vtk_encode_string(
   INPUT "${CMAKE_SOURCE_DIR}/resources/logo32.png"
   NAME F3DIcon
-  HEADER_OUTPUT F3D_ICON_HEADER
-  SOURCE_OUTPUT F3D_ICON_SOURCE
-  BINARY
-  NUL_TERMINATE)
+  BINARY)
```
Comment (mwestphal) [inline src/library/CMakeLists.txt:5]: not needed anymore ?

File: src/library/CMakeLists.txt [hunk 3]
```diff
@@ -59,89 +56,80 @@ if(F3D_MODULE_ALEMBIC)
...
-  ${CMAKE_CURRENT_BINARY_DIR}/readers
-  ${MODULES_SPECIFIC_INCLUDE_DIRS}
+target_include_directories(libf3d
+  PUBLIC
+  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}>
+  $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>
+  PRIVATE
+  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/readers>
+  $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/readers>
+  $<BUILD_INTERFACE:${MODULES_SPECIFIC_INCLUDE_DIRS}>
```
Comment (mwestphal) [inline src/library/CMakeLists.txt:100]: what is BUILD_INTERFACE ?

File: src/library/CMakeLists.txt [hunk 4]
```diff
@@ -59,89 +56,80 @@ if(F3D_MODULE_ALEMBIC)
...
-  PREFIX ""
+  LIBRARY_OUTPUT_NAME "f3d"
   )
 
+target_compile_features(libf3d PUBLIC cxx_std_11)
+
+if(F3D_INSTALL_SDK)
+  target_include_directories(libf3d
+    INTERFACE
+    $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
```
Comment (mwestphal) [inline src/library/CMakeLists.txt:115]: What is INSTALL_INTERFACE ?

File: src/application/main.cxx (hunk 1 above)
Comment (mwestphal) [reply src/application/main.cxx:15]: got it

File: src/library/F3DLoader.h [hunk 5]
```diff
@@ -67,22 +54,9 @@ class F3DLoader
...
-  F3DOptions CommandLineOptions;
-  F3DOptions Options;
-  F3DAnimationManager AnimationManager;
-  vtkSmartPointer<vtkF3DRenderer> Renderer;
-  vtkSmartPointer<vtkImporter> Importer;
-  vtkSmartPointer<vtkRenderWindow> RenWin;
-  F3DReaderInstantiator* ReaderInstantiator;
-
 private:
+  F3DLoaderInternals* Internals;
```
Comment (mwestphal) [inline src/library/F3DLoader.h:58]: std::unique_ptr

File: src/library/F3DLoader.h [hunk 6]
```diff
@@ -67,22 +54,9 @@ class F3DLoader
...
-  std::vector<std::string> FilesList;
-  int CurrentFileIndex = 0;
-  F3DOptionsParser Parser;
-  F3DOptions CommandLineOptions;
-  F3DOptions Options;
-  F3DAnimationManager AnimationManager;
-  vtkSmartPointer<vtkF3DRenderer> Renderer;
-  vtkSmartPointer<vtkImporter> Importer;
-  vtkSmartPointer<vtkRenderWindow> RenWin;
-  F3DReaderInstantiator* ReaderInstantiator;
```
Comment (mwestphal) [inline src/library/F3DLoader.h:83]: do you remember why this was a ptr ?

File: src/library/CMakeLists.txt (hunk 2 above)
Comment (Meakk) [reply src/library/CMakeLists.txt:5]: `NUL_TERMINATE` was useless, and other variable definition can be deduced.

File: src/library/CMakeLists.txt (hunk 3 above)
Comment (Meakk) [reply src/library/CMakeLists.txt:100]: https://stackoverflow.com/a/58729533/2609654

File: src/library/F3DLoader.h (hunk 6 above)
Comment (Meakk) [reply src/library/F3DLoader.h:83]: Idk

File: .github/test_sdk/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,8 @@
+project(test_sdk)
+
+cmake_minimum_required(VERSION 3.2)
+
+find_package(f3d REQUIRED)
+
+add_executable(test_sdk main.cxx)
```
Comment (mwestphal) [inline .github/test_sdk/CMakeLists.txt:7]: could this be part of the test suite instead ?

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -352,9 +352,9 @@ jobs:
         -DAlembic_DIR:PATH=$(pwd)/../dependencies/alembic_install/lib/cmake/Alembic
         -DImath_DIR:PATH=$(pwd)/../dependencies/imath_install/lib/cmake/Imath
         -DF3D_INSTALL_DEFAULT_CONFIGURATION_FILE=ON
-        ${{ matrix.os == 'windows-latest' && '-Ax64 -DF3D_WINDOWS_GUI=ON -DOpenCASCADE_DIR:PATH=$(pwd)/../dependencies/occt_install/cmake/ -DCMAKE_POLICY_DEFAULT_CMP0091=NEW -DCMAKE_MSVC_RUNTIME_LIBRARY=MultiThreaded' || null }}
+        ${{ matrix.os == 'windows-latest' && '-Ax64 -DF3D_INSTALL_SDK=ON -DF3D_WINDOWS_GUI=ON -DOpenCASCADE_DIR:PATH=$(pwd)/../dependencies/occt_install/cmake/ -DCMAKE_POLICY_DEFAULT_CMP0091=NEW -DCMAKE_MSVC_RUNTIME_LIBRARY=MultiThreaded' || null }}
         ${{ matrix.os == 'macos-latest' && '-DF3D_MACOS_BUNDLE=ON -DOpenCASCADE_DIR:PATH=$(pwd)/../dependencies/occt_install/lib/cmake/opencascade' || null }}
```
Comment (mwestphal) [inline .github/workflows/ci.yml:356]: pas de SDK macos ?

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -393,15 +393,22 @@ jobs:
       if: matrix.os == 'windows-latest'
       working-directory: ${{github.workspace}}/install
       run: |
-         cp ..\dependencies\mesa\x64\opengl32.dll .\bin\
-         cp ..\dependencies\mesa\x64\libglapi.dll .\bin\
-         .\bin\f3d.exe ..\source\testing\data\suzanne.obj --output=..\output.png
+        cp ..\dependencies\mesa\x64\opengl32.dll .\bin\
+        cp ..\dependencies\mesa\x64\libglapi.dll .\bin\
+        .\bin\f3d.exe ..\source\testing\data\suzanne.obj --output=..\output.png
```
Comment (mwestphal) [inline .github/workflows/ci.yml:398]: changed alignement

File: src/CMakeLists.txt [hunk 10]
```diff
@@ -1,6 +1,9 @@
 cmake_dependent_option(F3D_MACOS_BUNDLE "Build a macOS bundle application" ON "APPLE" OFF)
 cmake_dependent_option(F3D_WINDOWS_GUI "Build a non-console Win32 application" ON "WIN32" OFF)
 
+# Force static library when creating a macOS bundle
+cmake_dependent_option(BUILD_SHARED_LIBS "Build f3d with shared libraries" ON "NOT F3D_MACOS_BUNDLE" OFF)
```
Comment (mwestphal) [inline src/CMakeLists.txt:5]: how is this related to macos ?

File: src/application/CMakeLists.txt [hunk 11]
```diff
@@ -15,10 +15,9 @@ endif ()
 
 if(F3D_STRICT_BUILD)
   if(MSVC)
-    # Warning C4275 is essentially noise, disabling it to silent an issue with jsoncpp library
-    target_compile_options(f3d PUBLIC /W4 /WX /wd4275)
+    target_compile_options(f3d PRIVATE /W4 /WX)
```
Comment (mwestphal) [inline src/application/CMakeLists.txt:18]: not needed anymore ?

Comment (Meakk) [reply src/application/CMakeLists.txt:18]: Not for the executable

File: src/CMakeLists.txt (hunk 10 above)
Comment (Meakk) [reply src/CMakeLists.txt:5]: Static lib is required for macos bundle

File: .github/workflows/ci.yml (hunk 9 above)
Comment (Meakk) [reply .github/workflows/ci.yml:398]: Yes I fixed the indentation

File: .github/workflows/ci.yml (hunk 8 above)
Comment (Meakk) [reply .github/workflows/ci.yml:356]: It makes no sense with a bundle

File: .github/test_sdk/CMakeLists.txt (hunk 7 above)
Comment (Meakk) [reply .github/test_sdk/CMakeLists.txt:7]: No

Comment (mwestphal) [reply .github/test_sdk/CMakeLists.txt:7]: we will probably needs more tests of f3dlib directly. But we will adapt at that time.

**Review by mwestphal — APPROVED** [review]
